cpp/ql/lib/experimental/semmle/code/cpp/dataflow/ProductFlow.qll

```diff
@@ -324,25 +324,71 @@ module ProductFlow {
 
     module Flow2 = DataFlow::GlobalWithState<Config2>;
 
+    private predicate isSourcePair(Flow1::PathNode node1, Flow2::PathNode node2) {
+      Config::isSourcePair(node1.getNode(), node1.getState(), node2.getNode(), node2.getState())
+    }
+
+    private predicate isSinkPair(Flow1::PathNode node1, Flow2::PathNode node2) {
+      Config::isSinkPair(node1.getNode(), node1.getState(), node2.getNode(), node2.getState())
+    }
+
+    pragma[assume_small_delta]
     pragma[nomagic]
-    private predicate reachableInterprocEntry(
-      Flow1::PathNode source1, Flow2::PathNode source2, Flow1::PathNode node1, Flow2::PathNode node2
-    ) {
-      Config::isSourcePair(node1.getNode(), node1.getState(), node2.getNode(), node2.getState()) and
-      node1 = source1 and
-      node2 = source2
+    private predicate fwdReachableInterprocEntry(Flow1::PathNode node1, Flow2::PathNode node2) {
+      isSourcePair(node1, node2)
       or
-      exists(
-        Flow1::PathNode midEntry1, Flow2::PathNode midEntry2, Flow1::PathNode midExit1,
-        Flow2::PathNode midExit2
+      exists(Flow1::PathNode pred1, Flow2::PathNode pred2 |
+        fwdReachableInterprocEntry(pred1, pred2) and
+        isSuccessor(pred1, pred2, node1, node2)
+      )
+    }
+
+    bindingset[pred1, pred2]
+    bindingset[succ1, succ2]
+    private predicate isSuccessor(
```

Comment:
Replace with `fwdIsSuccessor` by pushing in the `fwdReachableInterprocEntry(pred1, pred2)` and make the two predicates mutually recursive as before.

Reply:
Fixed in 2458fa0ab3aa598306d4aad57fc0809d31dbc7d9.